### PR TITLE
Add checkpoints to blockchain

### PR DESCRIPTION
**Pull request by danield9tqh**
## Summary
Code implementation of [this discussion](https://discourse.ironfish.network/t/91)

## Testing Plan
Writing unit tests and simulations

## Documentation

Does this change require any updates to the Iron Fish Docs (ex. [the RPC API
Reference](https://ironfish.network/docs/onboarding/rpc/chain))? If yes, link a
related documentation pull request for the website.

```
[ ] Yes
```

## Breaking Change

Is this a breaking change? If yes, add notes below on why this is breaking and
what additional work is required, if any.

```
[ ] Yes
```

### CODE REVIEW

File: ironfish/src/consensus/consensus.ts [hunk 1]
```diff
@@ -61,18 +62,29 @@ export type ConsensusParameters = {
...
+  checkpoints: Checkpoint[]
 }
 
 export class Consensus {
   readonly parameters: ConsensusParameters
+  readonly checkpoints: Map<number, Buffer>
 
   constructor(parameters: ConsensusParameters) {
     this.parameters = parameters
+    this.checkpoints = new Map<number, Buffer>()
```
Comment (mat-if) [inline ironfish/src/consensus/consensus.ts:79]: Can you add some tests here around the new logic in the constructor and also the isActive function? Doesn't need to be too crazy, but its finnicky enough that it is liable to silently break in the future

File: ironfish/src/blockchain/blockchain.ts [hunk 2]
```diff
@@ -100,6 +100,8 @@ export class Blockchain {
 
   private _head: BlockHeader | null = null
 
+  private latestCheckpoint: BlockHeader | null = null
```
Comment (mat-if) [inline ironfish/src/blockchain/blockchain.ts:103]: Can you add some tests around this property and when it is set for sanity sake? I don't feel strongly about whether they need to be whole new tests or just added expects to the existing tests, up to you.

**Review by mat-if — COMMENTED** [review]
Looks good, just a few things around tests

File: ironfish/src/blockchain/blockchain.ts (hunk 2 above)
Comment (danield9tqh) [reply ironfish/src/blockchain/blockchain.ts]: Added asserts to existing tests for this

**Review by mat-if — APPROVED** [review]

**Pull request merged**